application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayService.java

```diff
@@ -0,0 +1,309 @@
+     *
+     * @param config the configuration for cake day management
+     * @param database the database for storing cake day information
+     */
+    public CakeDayService(Config config, Database database) {
+        this.config = config.getCakeDayConfig();
+        this.database = database;
+
+        this.cakeDayRolePredicate = Pattern.compile(this.config.rolePattern()).asPredicate();
+    }
+
+    private Optional<Role> getCakeDayRole(Guild guild) {
+        Optional<Role> cakeDayRole = getCakeDayRoleFromGuild(guild);
+
+        if (cakeDayRole.isEmpty()) {
+            logger.warn("Cake day role with pattern {} not found for guild: {}",
+                    config.rolePattern(), guild.getName());
+        }
+
+        return cakeDayRole;
+    }
+
+    /**
+     * Reassigns the cake day role for all members of the given guild.
+     * <p>
+     * If the cake day role is not found based on the configured pattern, a warning message is
+     * logged, and no action is taken.
+     *
+     * @param guild the guild for which to reassign the cake day role
+     */
+    protected void reassignCakeDayRole(Guild guild) {
+        Optional<Role> cakeDayRole = getCakeDayRole(guild);
+
+        if (cakeDayRole.isEmpty()) {
+            return;
+        }
+
+        refreshMembersCakeDayRoles(cakeDayRole.get(), guild);
+    }
+
+    /**
+     * Refreshes the Cake Day roles for members in the specified guild.
+     *
+     * @param cakeDayRole the Cake Day role to refresh
+     * @param guild the guild in which to refresh Cake Day roles
+     */
+    private void refreshMembersCakeDayRoles(Role cakeDayRole, Guild guild) {
+        guild.findMembersWithRoles(cakeDayRole).onSuccess(members -> {
+            removeRoleFromMembers(guild, cakeDayRole, members);
+            addTodayMembersCakeDayRole(guild);
+        });
+    }
+
+    /**
+     * Assigns a special role to members whose cake day (anniversary of joining) is today, but only
+     * if they have been a member for at least one year.
+     * <p>
+     * This method checks the current date against the cake day records in the database for each
+     * member of the given guild. If the member's cake day is today, and they have been a member for
+     * at least one year, the method assigns them a special role.
+     *
+     * @param guild the guild to check for members celebrating their cake day today
+     */
+    private void addTodayMembersCakeDayRole(Guild guild) {
+        findCakeDaysTodayFromDatabase(guild).forEach(cakeDayRecord -> {
+            Member member = guild.getMemberById(cakeDayRecord.getUserId());
+
+            if (member == null) {
+                return;
+            }
+
+            boolean isAnniversaryDay = hasMemberCakeDayToday(member);
+            int yearsSinceJoin = OffsetDateTime.now().getYear() - cakeDayRecord.getJoinedYear();
+
+            if (yearsSinceJoin > 0 && isAnniversaryDay) {
```

Comment:
Hm so this is confusing me a bit. `hasMemberCakeDayToday` checks if today is the cake day using the `Member member` **which includes** checking if the anniversary year is ` > 0`. This looks like checking the same thing twice (`yearsSinceJoin` calculation uses the db record though). Also is the member query needed for this check? Doesn't the database also store the join date (make sure it is up2date though)? You would only need the member instance fo adding the role later on, so you could move that `getMemberById` later in the chain and thus making it more efficient.